sdks/java/core/src/main/java/org/apache/beam/sdk/util/BeamWorkerInitializer.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.util.common.ReflectHelpers;
+
+/**
+ * A service interface for defining one-time initialization for Beam workers.
+ *
+ * <p>Beam workers will use {@code runOnStartup} and {@code runBeforeProcessing} to run every
+ * registered implementation's {@code onStartup} and {@code beforeProcessing} functions at the
+ * appropriate stage of execution.
+ *
+ * <p>{@link java.util.ServiceLoader} is used to discover implementations of {@link
+ * BeamWorkerInitializer}, note that you will need to register your implementation with the
+ * appropriate resources to ensure your code is executed. You can use a tool like {@link
+ * com.google.auto.service.AutoService} to automate this.
+ */
+@Experimental
+public abstract class BeamWorkerInitializer {
+  /**
+   * Finds all registered implementations of BeamWorkerInitializer and executes their {@code
+   * onStartup} methods. Called in worker harness implementations at the very beginning of their
+   * main method.
+   */
+  public static void runOnStartup() {
+    for (BeamWorkerInitializer initializer :
+        ReflectHelpers.loadServicesOrdered(BeamWorkerInitializer.class)) {
+      initializer.onStartup();
+    }
+  }
+
+  /**
+   * Finds all registered implementations of BeamWorkerInitializer and executes their {@code
+   * beforeProcessing} methods. Called in worker harness implementations after initialization but
+   * before beginning to process any data.
+   *
+   * @param options The pipeline options passed to the worker.
+   */
+  public static void runBeforeProcessing(PipelineOptions options) {
+    for (BeamWorkerInitializer initializer :
+        ReflectHelpers.loadServicesOrdered(BeamWorkerInitializer.class)) {
+      initializer.beforeProcessing(options);
+    }
+  }
+
+  /**
+   * Implement onStartup to run some custom initialization immediately after the worker begins
```

Comment:
I would move the methods you want people to implement to be at the top and hide the static methods in a different class inside the beam-fn-execution package.

Reply:
I moved the static methods to `org.apache.beam.sdk.fn.BeamWorkerInitializerHelpers` - is that what you had in mind?

Reply:
Yes.